barretenberg/cpp/src/barretenberg/vm/avm/trace/bytecode_trace.cpp

```diff
@@ -3,20 +3,60 @@
 #include "barretenberg/crypto/poseidon2/poseidon2_params.hpp"
 
 namespace bb::avm_trace {
+
+// These separators can be added to the constants file, but it's a bit finicky since there are the GENERATOR_INDEX ones
+const FF PARTIAL_ADDRESS = 27;
+const FF CONTRACT_ADDRESS_V1 = 15;
+const FF CONTRACT_LEAF = 16;
+const FF PUBLIC_KEYS_HASH = 52;
+
 using poseidon2 = crypto::Poseidon2<crypto::Poseidon2Bn254ScalarFieldParams>;
-AvmBytecodeTraceBuilder::AvmBytecodeTraceBuilder(const std::vector<AvmContractBytecode>& all_contracts_bytecode)
-    : all_contracts_bytecode(all_contracts_bytecode)
-{}
+using Point = grumpkin::g1::affine_element;
+
+/**************************************************************************************************
+ *                                   HELPERS
+ **************************************************************************************************/
+// TODO: Handling of point at infinity is still a bb <> ts pain point, we need to make sure we are consistent
+FF AvmBytecodeTraceBuilder::compute_address_from_instance(ContractInstanceHint contract_instance)
+{
+    FF salted_initialization_hash = poseidon2::hash({ PARTIAL_ADDRESS,
+                                                      contract_instance.salt,
+                                                      contract_instance.initialisation_hash,
+                                                      contract_instance.deployer_addr });
+    FF partial_address =
+        poseidon2::hash({ PARTIAL_ADDRESS, contract_instance.contract_class_id, salted_initialization_hash });
+    // Double check noir needs to hash infinity flag...
+    std::vector<FF> public_keys_hash_fields = contract_instance.public_keys.to_fields();
+    std::vector<FF> public_key_hash_vec{ PUBLIC_KEYS_HASH };
+    for (size_t i = 0; i < public_keys_hash_fields.size(); i += 2) {
+        public_key_hash_vec.push_back(public_keys_hash_fields[i]);
+        public_key_hash_vec.push_back(public_keys_hash_fields[i + 1]);
+        // Is it guaranteed we wont get a point at infinity here?
+        public_key_hash_vec.push_back(FF::zero());
+    }
+    FF public_keys_hash = poseidon2::hash({ public_key_hash_vec });
+    FF h = poseidon2::hash({ CONTRACT_ADDRESS_V1, public_keys_hash, partial_address });
+    // This is safe since BN254_Fr < GRUMPKIN_Fr so we know there is no modulo reduction
+    grumpkin::fr h_fq = grumpkin::fr(h);
+    return (grumpkin::g1::affine_one * h_fq + contract_instance.public_keys.incoming_viewing_key).x;
+}
+
+FF AvmBytecodeTraceBuilder::compute_contract_class_id(FF artifact_hash,
```

Comment:
make these `const& FF` and recheck all your functions, I saw that Jean caught others. We shouldn't copy fields unless needed